README.md

```diff
@@ -1,39 +1,120 @@
 # PSI Cardinality
 Private Set Intersection Cardinality protocol based on ECDH and Bloom Filters.
 
+## Requirements
+
+There are requirements for the entire project which each language shares. There also could be requirements for each target language:
+
+### Global Requirements
+These are the common requirements across all target languages of this project.
+
+- [Bazel](https://bazel.build)
+
+
+### Target Requirements
+Some target languages have additional requirements
+
+#### C++
+
+- None
+
+#### JavaScript
+
+- [NodeJS](https://nodejs.org/en/)
+- [Yarn](https://yarnpkg.com/)
+
+
 ## Compiling and Running
-The only requirement is [Bazel](https://bazel.build). To compile, run
+
+The repository uses a folder structure to isolate the supported targets from one another:
+
 ```
-bazel build //...
+src/<target language>/<sources>
 ```
-To run tests, use
+
+### C++
+
+Build all libraries (with all optimization levels), with a specific optimization level (-c opt), or a specific module
+
 ```
-bazel test //...
+# Build everything wth all optimization levels
+bazel build //src/cpp/...
+
+# With optimization flag '-c opt'
+bazel build -c opt //src/cpp/...
+
+# Specific module(s)
+bazel build -c opt //src/cpp:psi_cardinality_client
+bazel build -c opt //src/cpp:psi_cardinality_server
 ```
-Benchmarks can be run using
+
+Build and run tests
+
 ```
-bazel run -c opt //psi_cardinality:psi_cardinality_benchmark
+bazel test //src/cpp/...
 ```
 
+Build and run benchmarks
+
+```
+bazel run -c opt //src/cpp:psi_cardinality_benchmark
+```
+
+### JavaScript
+
+First, ensure you have updated submodules
+
+```
+yarn submodule:update
+```
+
+Then, update and initialize `emsdk`
+
+```
+yarn em:update
+yarn em:init
+```
+
+Next, build the WebAssembly, pure JS, or both variants
+
+```
+yarn build:wasm
+yarn build:js
+
+# or both
+yarn build
+```
+
+Finally, run the benchmark for WebAssembly or pure JS
+
+```
+yarn demo:wasm
+yarn demo:js
+```
+
+TODO: Build the client and server library wrapped with our JS abstraction.
+
 ## Using the Library
-To use this library in a Bazel project, put the following in your WORKSPACE file:
+To use this library in another Bazel project, add the following in your WORKSPACE file:
+
 ```
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 
 http_archive(
```

Comment:
This will only work for C++ for now. To make the JS version available to Bazel projects, we'd need to change the `http_archive` to a [`git_repository`](https://docs.bazel.build/versions/master/repo/git.html) rule, and move `
emsdk_configure(name = "emsdk")` to deps.bzl.